2021/day12/Day12.java

```diff
@@ -0,0 +1,109 @@
+import java.io.File;
+import java.util.*;
+
+public class Day12 {
+    static Map<String, ArrayList<String>> neighbourCaves = new HashMap<>();
+    static Set<String> paths = new HashSet<>();
+    static Set<String> pathsPartTwo = new HashSet<>();
+    static HashMap<String, Integer> visitedCaves = new HashMap<>();
+    static HashMap<String, Integer> visitedCavesPartTwo = new HashMap<>();
+
+    static boolean isSmallCave(String cave) {
+        return cave.equals(cave.toLowerCase());
+    }
+
+    static void readInput() {
+        Scanner sc;
+        try {
+            sc = new Scanner(new File("data.txt"));
+            while (sc.hasNextLine()) {
+                String[] caves = sc.nextLine().split("-");
+                String firstCave = caves[0];
+                String secondCave = caves[1];
+
+                if (neighbourCaves.containsKey(firstCave)) {
+                    neighbourCaves.get(firstCave).add(secondCave);
+                } else {
+                    neighbourCaves.put(firstCave, new ArrayList<>(List.of(secondCave)));
+                }
+
+                if (neighbourCaves.containsKey(secondCave)) {
+                    neighbourCaves.get(secondCave).add(firstCave);
+                } else {
+                    neighbourCaves.put(secondCave, new ArrayList<>(List.of(firstCave)));
+                }
+            }
+            sc.close();
+        } catch (Exception e) {
+            System.out.println("Something went horribly wrong: " + e);
+        }
+    }
+
+    static void findPath(String currentCave, String currentPath, HashMap<String, Integer> visitedCaves) {
+        currentPath += "," + currentCave;
+
+        if (currentCave.equals("end")) {
+            paths.add(currentPath);
+            return;
+        }
+
+        if (isSmallCave(currentCave)) {
+            visitedCaves.put(currentCave, 1);
+        }
+
+        for (String neighbourCave : neighbourCaves.get(currentCave)) {
+            if (!neighbourCave.equals("start") && !visitedCaves.containsKey(neighbourCave)) {
+                findPath(neighbourCave, currentPath, new HashMap<>(visitedCaves));
+            }
+        }
+    }
+
+    static void findPathPartTwo(String currentCave, String currentPath, HashMap<String, Integer> visitedCavesPartTwo, boolean wasAnyCaveVisitedTwice) {
```

Comment:
Again, I don't think you need the `HashMap` - a set would suffice. With a set, the lines 69-83 would look something like this:

```
if (isSmallCave(currentCave))
{
  if (visitedCavesPartTwo.contains(currentCave) && wasAnyCaveVisitedTwice)
    return
  else if (visitedCavesPartTwo.contains(currentCave)
    wasAnyCaveVisitedTwice = true;
  else
    visitedCavesPartTwo.put(currentCave);
}
```